// code/button_graphics.js
var $, window, document, win;
var settings = require("code/settings.js");

exports.draw = function(JQ, wind, doc, guiWin){
	$ = JQ;
	window = wind;
	document = doc;
	win = guiWin;
	
	$(window).scroll(function(e){
		$("#button_contain, #frame_select_contain").css("margin-left", 0-$(document).scrollLeft());
	});
	exports.line_setting = "arrow";
	
	var bxo = /[ball|o|x][1-5]/;
	
	$(".line_setting").click(function(){
		if(!$(this).hasClass("button_select")){
			$(".button_select").css("opacity", "0.5");
			$(".button_select").removeClass("button_select");
			$(this).addClass("button_select");
			exports.line_setting = $(this).attr("id");
			if(bxo.test(exports.line_setting) === false){
				$(".numbered_players").hide();
			}
		}
	});
	
	$(".button, #arrow").on("mouseenter", function(e){
		$(this).fadeTo(100, 1);
		var testbxo = bxo.test(exports.line_setting);
		if(bxo.test($(this).attr("id")) === false && testbxo === false){
			$(".numbered_players").hide();
		}
		else if(testbxo === true && bxo.test($(this).attr("id")) === false){
			$(".numbered_players").hide();
			if(bxo.test(exports.line_setting) === true){
				var sel = "#numbered_" + exports.line_setting.match(/ball|x|o/);
				
				$(sel).show();
				$(sel).children().show();
			}
		}
	});
	$(".line_setting, #arrow, .court_setting, .bound_setting, #settings .button, #edit_teams table .button").mouseleave(function(){
		if((!$(this).hasClass("button_select") && $(this).hasClass("line_setting")) || 
			(!$(this).hasClass("court_select") && $(this).hasClass("court_setting")) || 
			(!$(this).hasClass("bound_select") && $(this).hasClass("bound_setting")) || 
			(!$(this).hasClass("setting_select") && $(this).parents("#settings").length) ||
			(!$(this).hasClass("team_player_select") && $(this).parents("#edit_teams").length)
		){
			$(this).fadeTo(100, 0.5);
		}
	});
	$("#add_frame").mouseleave(function(){
		$(this).fadeTo(100, 0.5);
	});
	
	$(".non-radio").on("mouseleave", function(e){
		$(this).fadeTo(100, 0.5);
	});
	$(".non-radio").on("touchstart", function(){
		$(this).fadeTo(100, 1);
	});
	$(".non-radio").on("touchend", function(e){
		$(this).one("mouseenter", function(){
			$(this).mouseleave();
		});
	});
	
	$("body").on("focus", "input, textarea, option", function(){
		$(this).css("outline-color", "#1B4CAD");
	});
	$("body").on("focusout", "input, option", function(){
		$(this).css("outline-color", "");
	});
	$("body").on("focusout", "textarea", function(){
		$(this).css("outline-color", "");
	});
	
	$("body").on("focus", ".ui-dialog .ui-dialog-buttonpane button, .ui-dialog .ui-dialog-titlebar-close", function(){
		$(this).css("outline-color", "#1B4CAD");
	});
	
	$("body").on("focus", ".ui-tabs-active, .ui-tabs-anchor", function(){
		$(this).css("outline-color", "rgba(0,0,0,0)");
	});
	
	$("input").css("-webkit-user-select", "text");
	
	
	$(".court_setting").click(function(){
		if(!$(this).hasClass("court_select")){
			$(".court_select").css("opacity", "0.5");
			$(".court_select").removeClass("court_select");
			$(this).addClass("court_select");
		}
	});
	
	$(".bound_setting").click(function(){
		if(!$(this).hasClass("bound_select")){
			$(".bound_select").css("opacity", "0.5");
			$(".bound_select").removeClass("bound_select");
			$(this).addClass("bound_select");
		}
	});
	
	$("#edit_teams table .button").click(function(){
		if(!$(this).hasClass("team_player_select")){
			$(".team_player_select").css("opacity", "0.5");
			$(".team_player_select").removeClass("team_player_select");
			$(this).addClass("team_player_select");
		}
	});
	
	
	$("#frame_select").on("mouseenter", ".frame", function(){
		
		if(this.offsetWidth < this.scrollWidth){
			$(this).attr("title", $("#" + $(this).attr("id") + " span").text());
		}
		
	});
	
	$("#ball, #o, #x").on("mouseenter", function(){
		$(".numbered_players").hide();
		
		var sel = "#numbered_" + $(this).attr("id");
		$(sel).show();
		$(sel).children().show();
	});
	
	$("#ball, #o, #x").on("touchend", function(e){
		$(".numbered_players").hide();
		
		var sel = "#numbered_" + $(this).attr("id");
		$(sel).show();
		$(sel).children().show();
		
		e.stopPropagation();
	});
	
	$(".numbered_players, #button_contain").mouseleave(function(){
		$(".numbered_players").hide();
		if(bxo.test(exports.line_setting) === true){
			var sel = "#numbered_" + exports.line_setting.match(/ball|x|o/);
			
			$(sel).show();
			$(sel).children().show();
		}
	});
	
	$("#text_box_contain").on("mouseenter, focus", ".text_box", function(){
		$(this).children().first().show();
	});
	$("#text_box_contain").on("mouseleave, focusout", ".text_box", function(){
		$(this).children().first().hide();
	});
	$("#text_box_contain").on("touchend", ".text_box", function(){
		$(this).children("textarea").focus();
	});
	
	$("#drawing_contain").on("mousedown", ".drawing", function(e){
		if(!(e.ctrlKey || e.shiftKey)){
			$(".drawing").removeClass("drawing_select");
			$(".handle_display").removeClass("handle_select");
		}
		$(this).addClass("drawing_select");
	});
	$("#drawing_contain, #handle_contain").on("focusout", ".drawing, .handle", function(){
		$(".drawing").removeClass("drawing_select");
		$(".handle").removeClass("handle_select");
	});
	
	$("#handle_contain").on("mouseenter", ".handle", function(){
		$(".handle_display").addClass("handle_show");
	});
	$("#handle_contain").on("touchstart", ".handle", function(){
		$(".handle_display").addClass("handle_show");
	});
	$("#handle_contain").on("mouseleave", ".handle", function(){
		$(".handle_display").removeClass("handle_show");
	});
	$("#handle_contain").on("mouseenter", ".handle_display", function(){
		$(this).addClass("handle_show");
	});
	
	$("#handle_contain").on("mousedown", ".handle_display", function(e){
		if(!(e.ctrlKey || e.shiftKey)){
			$(".handle_display").removeClass("handle_select");
			$(".drawing").removeClass("drawing_select");
		}
		$(this).addClass("handle_select");
		global.draw = false;
	});
	
	$("#edit_players ol").on("mouseenter", "li>span", function(){
		$(this).fadeTo(100, 1);
	});
	$("#edit_players ol").on("mouseleave", "li>span", function(){
		$(this).fadeTo(100, 0.5);
	});
	
	$("#frame_select_contain, #canvas_contain").css("height", String(win.height - 130) + "px");
	$("#frame_select").css("height", String(win.height - 192) + "px");
	$("body").css("height", String(win.height - 112) + "px");
	
	win.on("resize", function(){
		$("#frame_select_contain, #canvas_contain").css("height", String(win.height - 130) + "px");
		$("#frame_select").css("height", String(win.height - 192) + "px");
		$("body").css("height", String(win.height - 112) + "px");
	});
	
	//settings window
	$("#settings .button").click(function(){
		$(this).parent().children().removeClass("setting_select").css("opacity", "0.5");
		$(this).addClass("setting_select").css("opacity", "1");
	});
};

exports.pdf = function(){
	exports.print_friend = false;
	exports.selection_only = false;
	$("#pdf_detail #per_page .button").mouseleave(function(){
		if(!$(this).hasClass("per_page_button_select")){
			$(this).fadeTo(100, 0.5);
		}
	});
	$("#pdf_detail #per_page .button").click(function(){
		if(!$(this).hasClass("per_page_button_select")){
			$(".per_page_button_select").css("opacity", "0.5");
			$(".per_page_button_select").removeClass("per_page_button_select");
			$(this).addClass("per_page_button_select");
			exports.per_page = parseInt($("#pdf_detail #per_page .per_page_button_select").attr("id").substring(9), 10);
			
			var src;
			if(exports.print_friend === false){
				src = "resources/images/pdf_examples/" + exports.per_page + ".png";
			}
			else{
				src = "resources/images/pdf_examples/" + exports.per_page + "_friend.png";
			}
			
			$("#example img").attr("src", src);
		}
	});
	
	$("#print_friend div").click(function(){
		if(!$(this).hasClass("single_select")){
			$(this).addClass("single_select");
			exports.print_friend = true;
		}
		else{
			$(this).removeClass("single_select");
			exports.print_friend = false;
		}
		
		var src;
		if(exports.print_friend === false){
			src = "resources/images/pdf_examples/" + exports.per_page + ".png";
		}
		else{
			src = "resources/images/pdf_examples/" + exports.per_page + "_friend.png";
		}
		
		$("#example img").attr("src", src);
	});
	
	$("#selection_only div").click(function(){
		if(!$(this).hasClass("single_select")){
			$(this).addClass("single_select");
			exports.selection_only = true;
		}
		else{
			$(this).removeClass("single_select");
			exports.selection_only = false;
		}
	});
	
	$("#print_friend div, #selection_only div").mouseleave(function(){
		if(!$(this).hasClass("single_select")){
			$(this).fadeTo(100, 0.5);
		}
	});
	
};

exports.print = function(){
	exports.print_friend_print = false;
	exports.selection_only_print = false;
	
	$("#print_detail #per_page_print .button").mouseleave(function(){
		if(!$(this).hasClass("per_page_button_select_print")){
			$(this).fadeTo(100, 0.5);
		}
	});
	$("#print_detail #per_page_print .button").click(function(){
		if(!$(this).hasClass("per_page_button_select_print")){
			$(".per_page_button_select_print").css("opacity", "0.5");
			$(".per_page_button_select_print").removeClass("per_page_button_select_print");
			$(this).addClass("per_page_button_select_print");
			exports.per_page_print = parseInt($("#print_detail #per_page_print .per_page_button_select_print").attr("id").substring(9), 10);
			
			var src_print;
			if(exports.print_friend_print === false){
				src_print = "resources/images/pdf_examples/" + exports.per_page_print + ".png";
			}
			else{
				src_print = "resources/images/pdf_examples/" + exports.per_page_print + "_friend.png";
			}
			
			$("#example_print img").attr("src", src_print);
		}
	});
	
	$("#print_friend_print div").click(function(){
		if(!$(this).hasClass("single_select")){
			$(this).addClass("single_select");
			exports.print_friend_print = true;
		}
		else{
			$(this).removeClass("single_select");
			exports.print_friend_print = false;
		}
		
		var src_print;
		if(exports.print_friend_print === false){
			src_print = "resources/images/pdf_examples/" + exports.per_page_print + ".png";
		}
		else{
			src_print = "resources/images/pdf_examples/" + exports.per_page_print + "_friend.png";
		}
		
		$("#example_print img").attr("src", src_print);
	});
	
	$("#selection_only_print div").click(function(){
		if(!$(this).hasClass("single_select")){
			$(this).addClass("single_select");
			exports.selection_only_print = true;
		}
		else{
			$(this).removeClass("single_select");
			exports.selection_only_print = false;
		}
	});
	
	$("#print_friend_print div, #selection_only_print div").mouseleave(function(){
		if(!$(this).hasClass("single_select")){
			$(this).fadeTo(100, 0.5);
		}
	});
	
	$("#per_page_1_print").mouseenter().click();
};
