// code/court_settings.js
var construct = require("code/construct.js");

exports.run = function(window, document, $, project, frame, calc, draw, misc, text_box){

	function court_settings(VAR, setting){
		
		global.actions.push(new construct.action({
			type: "court_setting",
			frame_index: frame.index,
			VAR: VAR,
			to:setting
		}));
		
		if(VAR === "full_court"){
			project.frames[frame.index].full_court = setting;
			project.frames[frame.index].line_arr = calc.translate(project.frames[frame.index].line_arr, setting, project.frames[frame.index].outside_lines);
			project.frames[frame.index].text_boxes = calc.translate_text_boxes(project.frames[frame.index].text_boxes, setting);
			text_box.reset();
		}
		else if(VAR === "outside_lines"){
			project.frames[frame.index].outside_lines = setting;
			project.frames[frame.index].line_arr = calc.scale(project.frames[frame.index].line_arr, setting, project.frames[frame.index].full_court);
		}
		
		
		var canvas = document.getElementById("canvas");
		var context = canvas.getContext("2d");
		
		draw.current_frame(frame.index, context, project.frames[frame.index], document);
		
		draw.log(context, document, project.frames[frame.index]);
		
		misc.saved = false;
	}
	
	$("#full_court").click(function(){
		if(project.frames[frame.index].full_court === false){
			court_settings("full_court", true);	
		}
	});
	
	$("#half_court").click(function(){
		if(project.frames[frame.index].full_court === true){
			court_settings("full_court", false);
		}
	});
	$("#no_bound").click(function(){
		if(project.frames[frame.index].outside_lines === true){
			court_settings("outside_lines", false);
		}
	});
	$("#bound").click(function(){
		if(project.frames[frame.index].outside_lines === false){
			court_settings("outside_lines", true);
		}
	});
	
}
