// code/pdf.js
var window, $, document, jsPDF;

var fs = require("fs"),
	draw = require("code/draw.js");

exports.run = function(win, JQ, doc){
	window = win;
	$ = JQ;
	document = doc;
	
	jsPDF = window.jsPDF;
	
};

exports.generate = function(project, per_page, print_friend, selection_only){
	
	var doc = new jsPDF("portrait", "in");
	
	if(selection_only === true){
		$(".frame_selected").each(function(){
			var index = index = parseInt($(this).attr("id").substring(6), 10);
			
			project.frames[index].sel = true;
		});
	}
	
	if(project.path !== ""){
		var path_arr = project.path.split(/[\/\\]/);
		project.title = path_arr[path_arr.length - 1].substring(0, path_arr[path_arr.length - 1].length - 5);
	}
	else{
		project.title = "New Project";
	}
	
	doc.setProperties({
		title: project.title,
		creator: "PlayWrite",
		author: "The coach",
		subject: "Basketball",
		keywords: project.title + ", PlayWrite, Basketball"
	});
	
	doc.setFontSize(12);
	doc.text(0.3, 0.3, project.title);
	var canvas = document.getElementById("pdf_canvas");
	var context = canvas.getContext("2d");
	
	
	var room = 0;
	
	
	if(per_page === 1){
		var metrics = {
			court:{
				top:{
					a: 0.8,
					b: 6.4
				},
				left: 1.148809,
				width:{
					full:5.95238,
					half:5.95238
				},
				height:{
					full:10,
					half:5
				}
			},
			name:{
				top:{
					a: 0.6,
					b: 6.2
				},
				left: 0.5
			}
		};
		room = 2;
		var page_number = 1;
		
		project.frames.forEach(function(f, index, arr){
		
			if(project.frames[index].sel === true || selection_only === false){
			
				delete project.frames[index].sel;
			
				draw.log(context, document, project.frames[index], print_friend, true);
				var img = canvas.toDataURL("image/jpeg");
				if(room === 0){
					doc.addPage();
					doc.text(0.3, 0.3, project.title);
					room = 2;
					page_number++;
				}
				if(room === 2){
					doc.text(7.5, 0.3, String(page_number));
				
					doc.text(metrics.name.left, metrics.name.top.a, project.frames[index].name);
					if(project.frames[index].full_court){
						doc.addImage(img, "JPEG", metrics.court.left, metrics.court.top.a, metrics.court.width.full, metrics.court.height.full);
						room = 0;
					}
					else{
						doc.addImage(img, "JPEG", metrics.court.left, metrics.court.top.a, metrics.court.width.half, metrics.court.height.half);
						room = 1;
					}
				}
				else{
					if(project.frames[index].full_court){
						doc.addPage();
						doc.text(0.3, 0.3, project.title);
						doc.text(metrics.name.left, metrics.name.top.a, project.frames[index].name);
						doc.addImage(img, "JPEG", metrics.court.left, metrics.court.top.a, metrics.court.width.full, metrics.court.height.full);
						room = 0;
						page_number++;
						
					}
					else{
						doc.text(metrics.name.left, metrics.name.top.b, project.frames[index].name);
						doc.addImage(img, "JPEG", metrics.court.left, metrics.court.top.b, metrics.court.width.half, metrics.court.height.half);
						room = 0;
					}
				}
			}
		});
		
		
	}
	else{
	
		var metrics = {
			court:{
				top:{
					a: 0.84,
					b: 4.28,
					c: 7.72
				},
				left: {
					a: 0.5,
					b: 4.25
				},
				width:{
					full:3.5,
					half:3.5
				},
				height:{
					full:5.88,
					half:2.94
				}
			},
			name:{
				top:{
					a: 0.69,
					b: 4.13,
					c: 7.57
				},
				left: {
					a: 0.5,
					b: 4.5
				}
			}
		};
		
		room = 6;
		var page_number = 0;
		
		project.frames.forEach(function(f, index, arr){
			
		
			if(project.frames[index].sel === true || selection_only === false){
				delete project.frames[index].sel;
				
				draw.log(context, document, project.frames[index], print_friend, true);
				var img = canvas.toDataURL("image/jpeg");
				
				//add a page
				if(room === 0){
					doc.addPage();
					doc.text(0.3, 0.3, project.title);
					room = 6;
					
				}
				
				
				
				switch(room){
					case 6:
						page_number++;
						doc.text(7.7, 0.3, String(page_number));
						
						doc.text(metrics.name.left.a, metrics.name.top.a, project.frames[index].name);
						if(project.frames[index].full_court){
							doc.addImage(img, "JPEG", metrics.court.left.a, metrics.court.top.a, metrics.court.width.full, metrics.court.height.full);
							room = 4;
						}
						else{
							doc.addImage(img, "JPEG", metrics.court.left.a, metrics.court.top.a, metrics.court.width.half, metrics.court.height.half);
							room = 5;
						}
					break;
					case 5:
						if(project.frames[index].full_court){
							//PDF one half court, then full court in 6 HC/page mis placed full court bug fixed
							doc.text(metrics.name.left.a, metrics.name.top.b, project.frames[index].name);
							doc.addImage(img, "JPEG", metrics.court.left.a, metrics.court.top.b, metrics.court.width.full, metrics.court.height.full);
							room = 3;
						}
						else{
							doc.text(metrics.name.left.a, metrics.name.top.b, project.frames[index].name);
							doc.addImage(img, "JPEG", metrics.court.left.a, metrics.court.top.b, metrics.court.width.half, metrics.court.height.half);
							room = 4;
						}
					break;
					case 4:
						if(project.frames[index].full_court){
							doc.setLineWidth(0.01);
							doc.line(4.125, metrics.court.top.a, 4.125, 10.66);
							doc.text(metrics.name.left.b, metrics.name.top.a, project.frames[index].name);
							doc.addImage(img, "JPEG", metrics.court.left.b, metrics.court.top.a, metrics.court.width.full, metrics.court.height.full);
							room = 1;
						}
						else{
							doc.text(metrics.name.left.a, metrics.name.top.c, project.frames[index].name);
							doc.addImage(img, "JPEG", metrics.court.left.a, metrics.court.top.c, metrics.court.width.half, metrics.court.height.half);
							room = 3;
						}
					break;
					case 3:
						if(project.frames[index].full_court){
							doc.text(metrics.name.left.b, metrics.name.top.a, project.frames[index].name);
							doc.addImage(img, "JPEG", metrics.court.left.b, metrics.court.top.a, metrics.court.width.full, metrics.court.height.full);
							room = 1;
							
						}
						else{
							doc.setLineWidth(0.01);
							doc.line(4.125, metrics.court.top.a, 4.125, 10.66);
							doc.text(metrics.name.left.b, metrics.name.top.a, project.frames[index].name);
							doc.addImage(img, "JPEG", metrics.court.left.b, metrics.court.top.a, metrics.court.width.half, metrics.court.height.half);
							room = 2;
						}
					break;
					case 2:
						if(project.frames[index].full_court){
							doc.text(metrics.name.left.b, metrics.name.top.b, project.frames[index].name);
							doc.addImage(img, "JPEG", metrics.court.left.b, metrics.court.top.b, metrics.court.width.full, metrics.court.height.full);
							room = 0;
						}
						else{
							doc.text(metrics.name.left.b, metrics.name.top.b, project.frames[index].name);
							doc.addImage(img, "JPEG", metrics.court.left.b, metrics.court.top.b, metrics.court.width.half, metrics.court.height.half);
							room = 1;
						}
					break;
					case 1:
						if(project.frames[index].full_court){
							doc.addPage();
							doc.text(0.3, 0.3, project.title);
							page_number++;
							doc.text(7.7, 0.3, String(page_number));
							doc.text(metrics.name.left.a, metrics.name.top.a, project.frames[index].name);
							doc.addImage(img, "JPEG", metrics.court.left.a, metrics.court.top.a, metrics.court.width.full, metrics.court.height.full);
							room = 4;
						}
						else{
							doc.text(metrics.name.left.b, metrics.name.top.c, project.frames[index].name);
							doc.addImage(img, "JPEG", metrics.court.left.b, metrics.court.top.c, metrics.court.width.half, metrics.court.height.half);
							room = 0;
						}
					break;
				}
				console.log(room, f);
			}
		});
	}
	
	var string = doc.output("datauristring");
	
	var data = string.substring(28);
	
	var buffer = new Buffer(data, 'base64');
	
	function savePDF(){
		var files = [$("#pdf_input").val()];
			
		if(files[0].substring(files[0].length - 4, files[0].length).toLowerCase() !== ".pdf"){
			files[0] += ".pdf";
		}
		
		fs.writeFileSync(files[0], buffer);
		$("#pdf_detail").dialog("close");
		
		$("#pdf_input").unbind("change");
		$("#pdf_input").remove();
		$("#button_contain").append('<input class="hidden" id="pdf_input" type="file" accept=".pdf" nwsaveas>');
		$("#button_contain").on("change", "#pdf_detail", savePDF);
	}
	
	$("#pdf_input").click();
	
	$("#pdf_input").change(savePDF);
	
};
